File: fibonacciPrime.py
# RUN WITH PYTHON 3

# FULL GITHUB REPO AT https://github.com/SolDroid/DSS-Fibonacci-Prime, 
# please visit this to see the full project

import numpy
import time

# Defining constants for Binet's formula
ROOT5 = numpy.sqrt(5)
PHI = (1+ROOT5) / 2
LOGPHI = numpy.log(PHI)

def find_fibonacci_primes(arr):
    outArr = []
    for value in arr:
        # Check for basic cases (README section 2.1)
        if (value <= 3): 
            if(value != 1):
                outArr.append(value)
            else:
                continue

        # Find Fibonacci index (README section 2.2)
        fibIndex = fibonacci_index(value)

        # Skip value if it doesn't have a valid Fibonacci index (not a Fibonacci number) (README section 2.2)
        if (fibIndex == -1): continue

        # Checks if the index is a prime number as all prime Fibonacci numbers will have a prime index, skips the value if not (README section 2.3)
        if (not is_prime(fibIndex)): continue

        # Finally, checks if the value is prime (README section 2.4)
        if (is_prime(value)): outArr.append(value)
    return outArr

def fibonacci_index(value):
    # Find the index of the nearest Fibonacci number using the inverse of an approximation of Binet's formula (README section 2.2)
    index = int(numpy.round(numpy.log(ROOT5 * value) / LOGPHI))
    nearFib = int(numpy.round(numpy.power(PHI, index) / ROOT5))

    # Return the index if it corresponds to the actual input value, otherwise return -1
    if (nearFib == value):
        return index
    return -1

def is_prime(value):
    # Very basic but still has a big o of O(sqrt(n)), this should be sufficient for this application (README section 2.4)
    if (not value & 1): return False

    # Iterate through possible factors, limited to the square root of the value as dividing by any number less than this would result in a corresponding factor greater than sqrt(value)
    for i in range(4, int(numpy.sqrt(value)) + 1):
        if (value % i == 0):
            return False
    return True

# For basic testing of find_fibonacci_primes(arr)
while True:
    arr = input("Enter input list with desired separator(s) (enter E to exit): ")
    try:
        if (arr.upper() == "E"): break
        arr = [int(value) for value in arr.split(" ")]
    except:
        print("Bad Input")
        continue
    start_time = time.time()
    print("Fibonacci primes: ")
    print(find_fibonacci_primes(arr))
    print("Finished in " + str(time.time() - start_time))

# Just a little CubeSat
print("""           Thank
            you

       ____________
      |\___________\\
      |_\___________\\
      |\ |         \ |
      |\\\\|__________\|
      |\\\\|__________\|
      | \|__________\|
      |__|_________| |
      |\ | \   \   \ |
      |_\|__\___\___\|   ______________________
      |\ |         \ |\\ /                      \\
      |\\\\|__________\| |  REMOVE BEFORE FLIGHT |
      |_\|__________\|/ \______________________/
      '\ |         \ |
        \|__________\|
         '           '

           Solomon
           Killam""")

File: fibonacciPrime.c
// TO COMPILE: gcc -o fibonacciPrime fibonacciPrime.c -lm
// Only includes comments to explain major differences from python version

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#define MAX_ARR_SIZE 100

// Defining constants for Binet's formula
#define ROOT5 sqrtl(5)
#define PHI (ROOT5+1)/(2)
#define LOGPHI logl(PHI)

unsigned int fibonacci_index(unsigned int value) {
    unsigned int index, nearFib;
    index = roundl(logl(ROOT5 * value) / LOGPHI);
    nearFib = roundl(powl(PHI, index) / ROOT5);
    if (nearFib == value) {
        return index;
    } else {
        // 0 is only used as default return so unsigned can be used as the return type, an index of 0 would correspond to a value of 0 but this number is filtered out by the basic cases
        return 0;
    }
}

short int is_prime(unsigned int value) {
    if (!(value & 1)) return 0;
    for (unsigned int i = 3; i <= floor(sqrtf(value)); i++) {
        if (value % i == 0) {
            return 0;
        }
    }
    return 1;
}

unsigned int * find_fibonacci_primes(unsigned int arr[], unsigned int length) {
    unsigned int outlength = 0, *outArr = malloc(sizeof(unsigned int) * length);
    for (unsigned int i = 0; i < length; i++) {
        if (arr[i] <= 3) {
            if (arr[i] != 1) {
                outArr[outlength] = arr[i];
                outlength++;
            } else {
                continue;
            }
        }
        unsigned int fibIndex;
        fibIndex = fibonacci_index(arr[i]);
        if (fibIndex == 0) continue;
        if (!is_prime(fibIndex)) continue;
        if (is_prime(arr[i])) {
            outArr[outlength] = arr[i];
            outlength++;
        }
    }
    // 4 is used as the placeholder for the termination of the array in the output as 4 will never be returned otherwise
    outArr[outlength] = 4;
    // Ends the used section of the output array with 0 so that the actual 
    return outArr;
}

int main(void) {
    // Initialise relavant variables
    unsigned int arrSize, arr[MAX_ARR_SIZE];

    // Take array input
    printf("Enter input array size: ");
    scanf("%u", &arrSize);
    printf("Enter values: ");
    for (unsigned int i = 0; i < arrSize; i++) {
        scanf("%u", &arr[i]);
    }

    // Call find_fibonacci_primes function
    unsigned int * fibonacciPrimes = find_fibonacci_primes(arr, arrSize);
    printf("Fibonacci primes: ");
    for (unsigned int i = 0; i < MAX_ARR_SIZE; i++) {
        if (fibonacciPrimes[i] == 4) break; // 4 used as terminator as per find_fibonacci_primes function
        printf("%u, ", fibonacciPrimes[i]);
    }
    printf("\b\b \n");
    return 0;
}